e2e/vault/vault_test.go

```diff
@@ -0,0 +1,276 @@
+	for _, v := range missing {
+		version := v
+		g.Go(func() error {
+			return h.get(version)
+		})
+	}
+	if err := g.Wait(); err != nil {
+		h.t.Fatalf("failed getting versions: %v", err)
+	}
+
+	binaries := make(map[string]string, len(versions))
+	for _, v := range versions {
+		binaries[v] = filepath.Join(h.binDir, v)
+	}
+	return binaries
+}
+
+// createBinDir creates the binary directory
+func (h *harness) createBinDir() {
+	// Check if the directory exists, otherwise create it
+	f, err := os.Stat(h.binDir)
+	if err != nil && !os.IsNotExist(err) {
+		h.t.Fatalf("failed to stat directory: %v", err)
+	}
+
+	if f != nil && f.IsDir() {
+		return
+	} else if f != nil {
+		if err := os.RemoveAll(h.binDir); err != nil {
+			h.t.Fatalf("failed to remove file at directory path: %v", err)
+		}
+	}
+
+	// Create the directory
+	if err := os.Mkdir(h.binDir, 0700); err != nil {
+		h.t.Fatalf("failed to make directory: %v", err)
+	}
+	if err := os.Chmod(h.binDir, 0700); err != nil {
+		h.t.Fatalf("failed to chmod: %v", err)
+	}
+}
+
+// diff returns the binaries that must be downloaded
+func (h *harness) diff() (missing []string) {
+	files, err := ioutil.ReadDir(h.binDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return versions
+		}
+
+		h.t.Fatalf("failed to stat directory: %v", err)
+	}
+
+	// Build the set we need
+	missingSet := make(map[string]struct{}, len(versions))
+	for _, v := range versions {
+		missingSet[v] = struct{}{}
+	}
+
+	for _, f := range files {
+		delete(missingSet, f.Name())
+	}
+
+	for k := range missingSet {
+		missing = append(missing, k)
+	}
+
+	return missing
+}
+
+// get retrieves the given Vault binary
+func (h *harness) get(version string) error {
+	resp, err := http.Get(
+		fmt.Sprintf("https://releases.hashicorp.com/vault/%s/vault_%s_%s_%s.zip",
+			version, version, h.os, h.arch))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Wrap in an in-mem buffer
+	b := bytes.NewBuffer(nil)
+	io.Copy(b, resp.Body)
+	resp.Body.Close()
+
+	zreader, err := zip.NewReader(bytes.NewReader(b.Bytes()), resp.ContentLength)
+	if err != nil {
+		return err
+	}
+
+	if l := len(zreader.File); l != 1 {
+		return fmt.Errorf("unexpected number of files in zip: %v", l)
+	}
+
+	// Copy the file to its destination
+	file := filepath.Join(h.binDir, version)
+	out, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0777)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	zfile, err := zreader.File[0].Open()
+	if err != nil {
+		return fmt.Errorf("failed to open zip file: %v", err)
+	}
+
+	if _, err := io.Copy(out, zfile); err != nil {
+		return fmt.Errorf("failed to decompress file to destination: %v", err)
+	}
+
+	return nil
+}
+
+// TestVaultCompatibility tests compatibility across Vault versions
+func TestVaultCompatibility(t *testing.T) {
+	h := newHarness(t)
+	vaultBinaries := h.reconcile()
+
+	for version, vaultBin := range vaultBinaries {
+		t.Run(version, func(t *testing.T) {
+			vbin := vaultBin
```

Comment:
Doesn't this have to be the first line in the for loop -- outside the func?

Reply:
Yep good catch!